[AERIE-1955] feat: add datepicker component

duranb opened the pull request:
This resolves https://jira.jpl.nasa.gov/browse/AERIE-1955.

To test:
1. Create a new plan
2. Verify that clicking on start/end times of the plan form opens the datepicker
3. Verify that clicking away from the datepicker or pressing the ESC key closes it
4. Verify that clicking on a day in the datepicker populates the input with the correct DOY string
5. Verify that typing in the date manually still works
6. After creating a new plan, add an activity
7. Set a start time for the activity via the activity's start time datepicker
8. Verify that it works the same

duranb committed:
feat: add datepicker component
replace time inputs with datepicker
show non-zero values in duration when readonly

duranb committed:
revert unecessary datepicker replacement

duranb committed:
added clarification in comment

duranb committed:
fix incorrect datepicker input change event

camargo commented on src/components/ui/DatePicker/DatePicker.svelte:
```diff
@@ -0,0 +1,345 @@
+  currentDate.setUTCMilliseconds(0);
+  const currentYear = currentDate.getUTCFullYear();
+
+  export let dateString: string = '';
```
Can we move the `export`s under the `import`s. General ordering: imports, exports, consts, lets, reactive expressions, lifecycle functions, functions. At some point we need a style doc for these (currently) un-lintable things.

duranb replied:
You're ok with the exports being broken up a bit? The `currentDate` declaration can't be moved below because `min/maxDate` uses it.

camargo commented on src/utilities/time.ts:
```diff
@@ -199,3 +213,45 @@ export function getUnixEpochTime(doyTimestamp: string): number {
+ * Parses a date string (YYYY-MM-DDTHH:mm:ss) or DOY string (YYYY-DDDDTHH:mm:ss) into its separate components
+ */
+export function parseDateTime(dateString: string, numDecimals = 6): null | ParsedDoyString | ParsedDateString {
+  const matches = (dateString != null ? dateString : '').match(
```
Can this just be `(dateString ?? '').match(...)`

camargo commented on src/utilities/time.ts:
```diff
@@ -64,11 +64,11 @@ export function convertDurationStringToUs(durationString: string): number | neve
 /**
- * Get a number value in microseconds given a string duration of the format '2y 318d 6h 16m 19s 200ms 0us'.
+ * Get a formatted duration string given a duration in microseconds.
  * @example convertUsToDurationString(90577779200000) -> '2y 318d 6h 16m 19s 200ms 0us'
```
Since `includeZeros` is defaulted to `false` should this `@example` exclude `0us`?

camargo commented on src/types/time.d.ts:
```diff
@@ -0,0 +1,20 @@
+  year: number;
+};
+
+type ParsedDateString = {
```
I'm wondering if we should call this `ParsedYmdString` for Year Month Day? Just to try and squeeze a little more context from the name.

camargo commented on src/routes/plans/index.svelte:
```diff
@@ -132,7 +137,6 @@
       <svelte:fragment slot="body">
         <form on:submit|preventDefault={createPlan}>
           <AlertError class="m-2" error={$createPlanError} />
-
```
Keep newline

camargo commented on src/components/ui/DatePicker/Week.svelte:
```diff
@@ -0,0 +1,46 @@
+<script lang="ts">
+  import { createEventDispatcher } from 'svelte';
+
```
Remove newline

camargo commented on src/components/ui/DatePicker/Week.svelte:
```diff
@@ -0,0 +1,46 @@
+
+  const dispatch = createEventDispatcher();
+
+  let weekDateArray: Date[] = [];
```
Add newline

camargo commented on src/components/ui/DatePicker/DatePickerDropdown.svelte:
```diff
@@ -0,0 +1,80 @@
+  import { createEventDispatcher } from 'svelte';
+
+  export { className as class };
+  let className: string = '';
```
Move `let` down

camargo commented on src/components/ui/DatePicker/DatePickerDropdown.svelte:
```diff
@@ -0,0 +1,80 @@
+<div class={`date-picker-dropdown${className ? ` ${className}` : ''}`}>
+  <div class="display-value">
+    <span class="value">{displayValue}</span>
+    <select class="hidden" data-type="number" {value} tabindex="-1" on:change={onChange}>
```
So actually we can simplify `on:change={onChange}` to just `on:change` and we can remove the `onChange` function. You should still be able to catch the native DOM change event on `DatePickerDropdown`: https://svelte.dev/tutorial/dom-event-forwarding

duranb replied:
that's awesome

camargo commented on src/components/ui/DatePicker/Day.svelte:
```diff
@@ -0,0 +1,99 @@
+<script lang="ts">
+  import { getDoy } from '../../../utilities/time';
+
```
Remove newline

camargo commented on src/components/ui/DatePicker/Day.svelte:
```diff
@@ -0,0 +1,99 @@
+  let isSelected: boolean = false;
+  let isWithinBounds: boolean = true;
+
+  const dispatch = createEventDispatcher();
```
Move const up

camargo commented on src/components/ui/DatePicker/Day.svelte:
```diff
@@ -0,0 +1,99 @@
+</script>
+
+<div
+  class={`date-picker-day${isToday ? ' today' : ''}${isSelected ? ' selected' : ''}${
```
Not required but this might be more readable if this class is built in a reactive expression

duranb replied:
I can write a util function that basically does what the `classnames` library does. Unless we wanted to actually use `classnames`? I'm down for either.

camargo commented on src/components/ui/DatePicker/Month.svelte:
```diff
@@ -0,0 +1,65 @@
+<script lang="ts">
+  import { createEventDispatcher } from 'svelte';
+
```
Remove newline

camargo commented on src/components/ui/DatePicker/DatePicker.svelte:
```diff
@@ -0,0 +1,345 @@
+    on:change={autoCompleteDate}
+    on:click={openDatePicker}
+    on:focus={openDatePicker}
+    on:keydown={onInputKeydown}
```
You should be able to DOM event forward this as well and remove `onInputKeydown`

duranb committed:
forward events through Datepicker components
Move "Enter" key handling directly into Datepicker
add classNames function for more legible class string construction
fix documentation
cleanup code

duranb committed:
more cleanup

duranb committed:
more cleanup

duranb committed:
add classNames call to dropdown
rename parseDateTime to parseDoyOrYmdTime
more cleanup

duranb committed:
pass field value to Datepicker in DatePickerField

camargo reviewed (approved):
LGTM!
